### PR TITLE
Fixing $ref behaviour

### PR DESCRIPTION
**Pull request by chrusty**
A fairly substantial change here, and I believe it should actually replace the default behaviour.

A user recently raised [an issue](https://github.com/chrusty/protoc-gen-jsonschema/issues/74) where some nested messages were referenced, and others were encoded inline. An investigation uncovered some suspicious logic previously contributed by another user, and a fix has been implemented.

Nested messages are now all included in the `definitions` of a generated schema, and referenced appropriately.

- This prevents duplication
- This also makes it clear which proto message a particular nested payload was derived from

I'm quite happy with the fix, but of course the result will be that schemas previous generated will change. They should remain functionally the same, just a bit cleaner.

Changes to generated schemas
=======================

- All proto messages are added to the `definitions` section of generated JSONSchemas
- Any occurrence of a nested message is now linked to these definitions using a `$ref` (whereas before they _sometimes_ did this, but most of the time were included inline)
- Even the _root_ type will be referenced now (this means that even strange recursive schemas should generated fine, and you will always know the name of the proto message which generated a specific schema / definition)

Examples
=======

### Flat messages

This example shows how everything is now a $ref (even the root message). It is now obvious which proto message a schema was derived from because the ID and definition reference use the name from the protos.

Before
```json
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "properties": {
        "timestamp": {
            "type": "string",
            "format": "date-time"
        }
    },
    "additionalProperties": true,
    "type": "object"
}
```

After
```json
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "$ref": "#/definitions/Timestamp",
    "id": "Timestamp",
    "definitions": {
        "Timestamp": {
            "properties": {
                "timestamp": {
                    "type": "string",
                    "format": "date-time"
                }
            },
            "additionalProperties": true,
            "type": "object",
            "id": "Timestamp"
        }
    }
}
```

### Nested messages

This example continues with the "everything is now a reference" theme: not only does it apply to the root message, but any nested messages within. Now you might appreciate knowing the name of the original proto messages.

Before
```json
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "properties": {
        "bar": {
            "properties": {
                "foo": {
                    "type": "integer"
                }
            },
            "additionalProperties": true,
            "type": "object"
        },
        "baz": {
            "properties": {
                "foo": {
                    "type": "string"
                }
            },
            "additionalProperties": true,
            "type": "object"
        },
        "something": {
            "type": "boolean"
        }
    },
    "additionalProperties": true,
    "type": "object",
    "oneOf": [
        {
            "required": [
                "bar"
            ]
        },
        {
            "required": [
                "baz"
            ]
        }
    ]
}
```

After
```json
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "$ref": "#/definitions/OneOf",
    "id": "OneOf",
    "definitions": {
        "OneOf": {
            "properties": {
                "bar": {
                    "$ref": "#/definitions/samples.OneOf.Bar",
                    "additionalProperties": true
                },
                "baz": {
                    "$ref": "#/definitions/samples.OneOf.Baz",
                    "additionalProperties": true
                },
                "something": {
                    "type": "boolean"
                }
            },
            "additionalProperties": true,
            "type": "object",
            "oneOf": [
                {
                    "required": [
                        "bar"
                    ]
                },
                {
                    "required": [
                        "baz"
                    ]
                }
            ],
            "id": "OneOf"
        },
        "samples.OneOf.Bar": {
            "properties": {
                "foo": {
                    "type": "integer"
                }
            },
            "additionalProperties": true,
            "type": "object",
            "id": "samples.OneOf.Bar"
        },
        "samples.OneOf.Baz": {
            "properties": {
                "foo": {
                    "type": "string"
                }
            },
            "additionalProperties": true,
            "type": "object",
            "id": "samples.OneOf.Baz"
        }
    }
}
```

### JSONSchema Version

Previously any schemas in the `definitions` included a (redundant) JSONSchema version. Now this is only added to the root schema.

### CODE REVIEW

**Comment by chrusty** [comment]
@grant a heads up for you here. This is an example where I believe that the default behaviour needs to change, as a decent bug was discovered. The resulting schemas should prove to remain functionally the same, but definitely different in terms of structure. I'm hoping to release this as v1.1.1.

**Comment by chrusty** [comment]
I still need to re-generate the expected results for the unit tests (and go over them very carefully to make sure they ARE still correct).

**Comment by grant** [comment]
Thanks @chrusty. We can test these changes. I don't quite understand how we're not using `properties` and are using `definitions`. Can you provide some example before and after changes for a few JSON schemas in this PR description?

CC @grayside

**Comment by chrusty** [comment]
> Thanks @chrusty. We can test these changes. I don't quite understand how we're not using `properties` and are using `definitions`. Can you provide some example before and after changes for a few JSON schemas in this PR description?
> 
> CC @grayside

Sure thing @grant and @grayside. I should probably add some docs around this too.

Thanks for getting involved at this point. It is quite a major change, and I know that you guys rely on this thing.

**Review by grant — CHANGES_REQUESTED** [review]
Just marking this PR as ❌  while not ready so you can re-request a review (and give me a notification) when ready.

**Comment by chrusty** [comment]
@grant OK, I think this is ready now.

Effectively this change just makes every nested object / message into a reference to a definition (even the root object).

The benefits of this are:

- Removes duplication (each object is only defined once, and referenced as many times as needed)
- References now explicitly state the name of the proto message they refer to (some users wanted clarity on this)
- Handles recursive / cyclical references too

On top of this I've introduced some "expected to pass" and "expected to fail" payloads whichare validated against the generated schemas in the unit test suite. This gives me the confidence to say that the schemas behave the same as they previously did.

The schemas should behave exactly the same as they did previously, they just look a bit different now!

**Comment by grant** [comment]
Thanks. Here's the data for me to analyze for the JSON schema diff:

| Type | Before | After | 
| --- | --- | --- | 
| LogEntryData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/cloud/audit/v1/LogEntryData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/cloud/audit/v1/LogEntryData.json) | 
| BuildEventData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/cloud/cloudbuild/v1/BuildEventData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/cloud/cloudbuild/v1/BuildEventData.json) | 
| DocumentEventData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/cloud/firestore/v1/DocumentEventData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/cloud/firestore/v1/DocumentEventData.json) | 
| MessagePublishedData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/cloud/pubsub/v1/MessagePublishedData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/cloud/pubsub/v1/MessagePublishedData.json) | 
| SchedulerJobData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/cloud/scheduler/v1/SchedulerJobData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/cloud/scheduler/v1/SchedulerJobData.json) | 
| StorageObjectData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/cloud/storage/v1/StorageObjectData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/cloud/storage/v1/StorageObjectData.json) | 
| AnalyticsLogData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/firebase/analytics/v1/AnalyticsLogData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/firebase/analytics/v1/AnalyticsLogData.json) | 
| AuthEventData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/firebase/auth/v1/AuthEventData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/firebase/auth/v1/AuthEventData.json) | 
| ReferenceEventData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/firebase/database/v1/ReferenceEventData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/firebase/database/v1/ReferenceEventData.json) | 
| RemoteConfigEventData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/firebase/remoteconfig/v1/RemoteConfigEventData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/firebase/remoteconfig/v1/RemoteConfigEventData.json) | 
| TestMatrixEventData | [Before](https://googleapis.github.io/google-cloudevents/jsonschema/google/events/firebase/testlab/v1/TestMatrixEventData.json) | [After](https://grant.github.io/google-cloudevents/jsonschema/google/events/firebase/testlab/v1/TestMatrixEventData.json) | 

We do some post-processing on the JSON schema, so there might be some missing fields in the unmodified generated schemas.

I think the largest issue is that I'm not seeing this schema work when testing on https://app.quicktype.io/. Is it possible to have this feature flagged? I'm thinking we'd have to fork the repo if this was the default.

---

CC @grayside @jonskeet

**Comment by chrusty** [comment]
Hey @grant thanks for this explanation.

OK I tried your schemas in app.quicktype.io, and it looks like the names of the references are an issue there (it worked when I generated the references using a `#definitions/` prefix).

What If I could get this branch working with quicktype? Honestly I've had different behaviour from different schema validators, to the point where I'm not even sure any more! My instinct was initially to prefix them with `#definitions/`, but that broke the golang validator library i'm using in my unit test suite! Anyway I'll see if I can make that work.

Also is it worth having a talk about the post-processing you guys are doing? If it is down to an error or omission in this generator then it would be something worth fixing here, no?

CC @graysude @JonSkeet

**Comment by grant** [comment]
Ideally, schemas would work with quicktype or VS Code jsonschema integrations (i.e. app.quicktype.com produces nice Golang structs, [VS Code](https://code.visualstudio.com/docs/languages/json#_mapping-in-the-json) (i.e. providing `$schema` and seeing autocompletion).

If we had like 5 examples of before and after schemas (in the PR description), that would help with testing. I know the tests are fairly comprehensive and have these changes.

There may be ambiguities / differences with tooling / schemas that make a $ref-centric schema behave differently. Hopefully we can preserve the overall behavior.

I assume I'll have to change any post-processing on my side irrespective of this change.

File: internal/converter/testdata/array_of_objects.go
Comment (grant) [reply internal/converter/testdata/array_of_objects.go:8]: (I never published this comment, just adding it now!)

Comment (chrusty) [reply internal/converter/testdata/array_of_objects.go]: Good point @grant, I agree - this is pure cruft. I've fixed it up and now the schema version is only included with the root type.

File: jsonschemas/ArrayOfPrimitives.jsonschema
Comment (grant) [inline jsonschemas/ArrayOfPrimitives.jsonschema:4]: I'm seeing the "id" value repeated twice per schema, which doesn't make it a valid ID. As mentioned in the other comment, I would remove the top-level id as we're purely using $ref for defining top-level properties.

**Review by grant — CHANGES_REQUESTED** [review]
Thanks! I've tested this change with Golang Quicktype and don't see a functional change, which is good.

This approach seems generally beneficial and may help with multi-jsonschema opportunities in the future (we currently don't re-use types across json schemas or in generated libraries, maybe there's a solution).

I've commented some requested changes, but should approve after then.

File: internal/converter/testdata/array_of_enums.go
Comment (chrusty) [reply internal/converter/testdata/array_of_enums.go]: OK, I've completely removed the IDs. Tests are good, and i've tried the generated schemas out with a couple of different validators. Now that we're using $ref everywhere the users can still understand which proto message generated which schema definition, so I'm happy with the change.

File: jsonschemas/ArrayOfPrimitives.jsonschema
Comment (chrusty) [reply jsonschemas/ArrayOfPrimitives.jsonschema]: I've removed ID from the whole lot (top-level plus all definitions). Hopefully this is acceptable for your use case.

**Review by grant — APPROVED** [review]
LGTM, I think this will improve the schemas long-term, thanks!

File: jsonschemas/ArrayOfPrimitives.jsonschema
Comment (grant) [reply jsonschemas/ArrayOfPrimitives.jsonschema]: Thanks.

To be clear I am fine with IDs, but they must be valid to the draft-4 JSON schema specification if we should include them (for anyone's use case). The spec requires a scheme for the ID for example.

It's a little easier to add fields than remove fields.

https://datatracker.ietf.org/doc/html/rfc3986#section-1.1.1
https://datatracker.ietf.org/doc/html/draft-zyp-json-schema-04#section-7.2.1

**Comment by chrusty** [comment]
I'm fine to leave them out - the schemas seem to work, and I'm in favour of
"less is more".

**Pull request merged**
